### PR TITLE
Test against publisher v2 schema

### PR DESCRIPTION
**Pull request by boffbowsh**
When Whitehall was ported to use v2 of the Publishing API, it wasn't updated to perform validation against the V2 schemas. This lead to issues where the app was sending documents containing now-invalid properties like `content_id` and `update_type`, and not sending required properties such as `redirects`, `description` and `title`.

In order to fix this, it became necessary to refactor the `PublishingApiPresenters` classes and extract common functionality into a base `Item` presenter. These presenters now have an interface of explicit public methods that expose the `content_id`, the document for the `content` API call, the `links` hash and the `update_type`. The locale is still taken from the content document. This replaces the previous `as_json` method.

The only production code that has been updated is the presenters and the `PublishingApi*Worker` classes that use them to determine what to send to the Publishing API. However, many tests throughout the codebase needed to be updated to use this new format, because lots of things that pretend to be unit tests really aren't. At some point we should refactor these to have clearer testing boundaries, and stub the calls to `gds-api-adapters` rather than using WebMock. We should trust the pact testing within `gds-api-adapters` to cover what is actually being sent to the Publishing API. We should also be able to test a single type of presenter that conforms to the above interface, and then test each presenter against that interface. But that's for another day.

/cc @benilovj @jamiecobbett @elliotcm @gpeng

### CODE REVIEW

**Comment by benilovj** [comment]
This is all very sensible, :+1: from me but perhaps it's worth getting another opinion on this before we merge.

Regarding the testing strategy, as we discussed IRL, it'd be good to spike an alternative approach, using the Take Part tests as a concrete examples, and go from there...

File: app/presenters/publishing_api_presenters/case_study.rb [hunk 1]
```diff
@@ -1,43 +1,43 @@
...
+      supporting_organisations: item.supporting_organisations.map(&:content_id),
+      document_collections: item.published_document_collections.map(&:content_id),
+      world_locations: item.world_locations.map(&:content_id),
+      worldwide_organisations: item.worldwide_organisations.map(&:content_id),
+      worldwide_priorities: item.worldwide_priorities.map(&:content_id),
     }
   end
 
+private
+
```
Comment (benlovell) [inline app/presenters/publishing_api_presenters/case_study.rb]: Styleguide pedantry: our style dictates this line should be removed.

File: app/presenters/publishing_api_presenters/edition.rb [hunk 2]
```diff
@@ -1,59 +1,39 @@
...
 
-    def initialize(edition, options = {})
-      @edition = edition
-      @update_type = options[:update_type] || default_update_type
+    def links
+      {}
     end
 
+  private
+
```
Comment (benlovell) [inline app/presenters/publishing_api_presenters/edition.rb]: Same here. Soz.

File: app/presenters/publishing_api_presenters/item.rb [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+          content_hash.merge!(analytics_identifier: item.analytics_identifier)
+        end
+      end
+    end
+
+    def links
+      {}
+    end
+
+    private
```
Comment (benlovell) [inline app/presenters/publishing_api_presenters/item.rb]: This should be unindented.

File: test/unit/data_hygiene/organisation_reslugger_test.rb [hunk 4]
```diff
@@ -30,11 +30,16 @@ def teardown
...
       old_base_path = @organisation.search_link
       new_base_path = "#{base_path}/corrected-slug"
-      content_item[:base_path] = new_base_path
-      content_item[:routes][0][:path] = new_base_path
+
+      content = content_item.content
+      content[:base_path] = new_base_path
+      content[:routes][0][:path] = new_base_path
+
+      content_item.stubs(:content).returns(content)
```
Comment (benlovell) [inline test/unit/data_hygiene/organisation_reslugger_test.rb]: You could also pass the key/value to `stubs` here and drop the `returns` bit.

File: test/unit/data_hygiene/person_reslugger_test.rb [hunk 5]
```diff
@@ -26,19 +26,33 @@ class PersonSlugChangerTest < ActiveSupport::TestCase
...
     old_base_path = @person.search_link
     new_base_path = "/government/people/updated-slug"
-    content_item[:base_path] = new_base_path
-    content_item[:routes][0][:path] = new_base_path
+
+    content = content_item.content
+    content[:base_path] = new_base_path
+    content[:routes][0][:path] = new_base_path
+
+    content_item.stubs(:content).returns(content)
```
Comment (benlovell) [inline test/unit/data_hygiene/person_reslugger_test.rb]: Same here.

File: test/unit/data_hygiene/role_reslugger_test.rb [hunk 6]
```diff
@@ -24,19 +24,33 @@ class MinisterialRoleResluggerTest < ActiveSupport::TestCase
...
     old_base_path = @ministerial_role.search_link
     new_base_path = "/government/ministers/corrected-slug"
-    content_item[:base_path] = new_base_path
-    content_item[:routes][0][:path] = new_base_path
+
+    content = content_item.content
+    content[:base_path] = new_base_path
+    content[:routes][0][:path] = new_base_path
+
+    content_item.stubs(:content).returns(content)
```
Comment (benlovell) [inline test/unit/data_hygiene/role_reslugger_test.rb]: :speak_no_evil:

File: test/unit/presenters/publishing_api_presenters/case_study_test.rb [hunk 7]
```diff
@@ -135,18 +137,18 @@ def present(edition)
       worldwide_priorities: [],
     }
 
-    assert_valid_against_schema(presented_hash, 'case_study')
-    assert_equal expected_links_hash, presented_hash[:links]
+    assert_valid_against_links_schema({links: presented_item.links}, 'case_study')
```
Comment (benlovell) [inline test/unit/presenters/publishing_api_presenters/case_study_test.rb]: Style nitpick: spaces inside the hash.

File: test/unit/presenters/publishing_api_presenters/case_study_test.rb [hunk 8]
```diff
@@ -158,52 +160,52 @@ def present(edition)
     location = create(:world_location)
     case_study = create(:published_case_study,
                         world_locations: [location])
-    presented_hash = present(case_study)
-    assert_valid_against_schema(presented_hash, 'case_study')
-    assert_equal [location.content_id], presented_hash[:links][:world_locations]
+    presented_item = present(case_study)
+    assert_valid_against_links_schema({links: presented_item.links}, 'case_study')
```
Comment (benlovell) [inline test/unit/presenters/publishing_api_presenters/case_study_test.rb]: Same here.

File: test/unit/presenters/publishing_api_presenters/case_study_test.rb [hunk 9]
```diff
@@ -158,52 +160,52 @@ def present(edition)
...
 
   test "links hash includes worldwide organisations" do
     wworg = create(:worldwide_organisation)
     case_study = create(:published_case_study,
                         worldwide_organisations: [wworg])
-    presented_hash = present(case_study)
-    assert_valid_against_schema(presented_hash, 'case_study')
-    assert_equal [wworg.content_id], presented_hash[:links][:worldwide_organisations]
+    presented_item = present(case_study)
+    assert_valid_against_links_schema({links: presented_item.links}, 'case_study')
```
Comment (benlovell) [inline test/unit/presenters/publishing_api_presenters/case_study_test.rb]: Same.

**Comment by benlovell** [comment]
:+1: other than the minor style stuff.

**Comment by boffbowsh** [comment]
This will need changing to put `update_type` back. I'll fix the comments above, then close this and reopen.

**Comment by tommyp** [comment]
In relation to @boffbowsh's comment above:

Me and @boffbowsh had a chat in meatspace about `update_type` in the V2 world. As I continue to work on the Specialist Publisher rewrite I hit an issue with the Publishing API V2 only requiring `update_type` on `publish`. As new Apps will use the Publishing API as a data store (and not have one of their own) the `update_type` needs to be stored as it needs to be persisted between the `live > draft > draft > draft > live` cycle.

This means that `update_type` isn't really tied to the `publish` action anymore and instead can be sent as part of the payload. But because certain Apps, like Whitehall, will continue to have their own data store _for now_, this is going to need to be either provided from the existing draft payload _or_ on the `publish` action.

**Comment by elliotcm** [comment]
To follow up on @tommyp's comment, the intent here is that different publishing applications might ask for the update type and change note at different times.  Some might treat it as part of the editing process, others as part of the publishing process.

As such, publishing API will accept `update_type` with optional `change_note` BOTH during save AND during publish, with the publish contents taking preference if both have been provided.

This will also make change note optional in all cases as far as the API is concerned.  Restrictions like "major changes need a change note" should be handled in the publishing applications.

**Comment by boffbowsh** [comment]
Finally ready for review again.

**Comment by jamiecobbett** [comment]
Needs a rebase.

File: app/presenters/publishing_api_presenters/edition.rb [hunk 10]
```diff
@@ -1,59 +1,38 @@
-# For now, this is used to register dummy editions in the content store as
+# For now, this is used to register dummy items in the content store as
```
Comment (jamiecobbett) [inline app/presenters/publishing_api_presenters/edition.rb]: This comment isn't accurate - some subclasses are sending real content items. I think now is as good a time as any to reword it to say that.

File: app/presenters/publishing_api_presenters/item.rb [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+    def rendering_app
+      "whitehall-frontend"
+    end
+
+    def routes
+      [{ path: base_path, type: "exact" }]
+    end
+
+    def base_path
+      Whitehall.url_maker.polymorphic_path(item)
```
Comment (jamiecobbett) [inline app/presenters/publishing_api_presenters/item.rb]: I think this is fine and I don't have an alternative suggestion, but I have a feeling that this method doesn't work for all things that we will pass to content-store. That sounds really broad now that I've written it out, and I suppose it is.

File: app/presenters/publishing_api_presenters/case_study.rb [hunk 12]
```diff
@@ -1,43 +1,42 @@
...
+      supporting_organisations: item.supporting_organisations.map(&:content_id),
+      document_collections: item.published_document_collections.map(&:content_id),
+      world_locations: item.world_locations.map(&:content_id),
+      worldwide_organisations: item.worldwide_organisations.map(&:content_id),
+      worldwide_priorities: item.worldwide_priorities.map(&:content_id),
     }
   end
 
+private
+  def document_format
```
Comment (jamiecobbett) [inline app/presenters/publishing_api_presenters/case_study.rb]: Any particular reason this was changed from `format`?

Comment (boffbowsh) [inline app/presenters/publishing_api_presenters/case_study.rb]: Yes. While working on this I forgot to implement `format` on a class. This ended up calling `Kernel#format` which confused me for a while. I figured it was better to have a more obvious failure scenario by renaming the method and having a `NoMethodError` thrown if it wasn't implemented.

File: app/presenters/publishing_api_presenters/item.rb (hunk 11 above)
Comment (boffbowsh) [inline app/presenters/publishing_api_presenters/item.rb]: I think (hope) those cases will be covered by the unit tests for the individual formats.

File: app/presenters/publishing_api_presenters/case_study.rb (hunk 12 above)
Comment (jamiecobbett) [inline app/presenters/publishing_api_presenters/case_study.rb]: :+1:

**Comment by jamiecobbett** [comment]
Generally very happy with this :+1: I've reviewed the implementation, and that seems good, but haven't looked at the tests yet.

File: app/presenters/publishing_api_presenters/edition.rb (hunk 10 above)
Comment (boffbowsh) [inline app/presenters/publishing_api_presenters/edition.rb]: Updated.

File: app/presenters/publishing_api_presenters/item.rb [hunk 13]
```diff
@@ -0,0 +1,66 @@
+module PublishingApiPresenters
+  class Item
+    extend Forwardable
+
+    def_delegators :item, :base_path, :content_id, :title, :description, :need_ids, :public_updated_at
+    attr_accessor :update_type
+
+    def initialize(item, update_type: nil)
+      self.item = item
```
Comment (benlovell) [inline app/presenters/publishing_api_presenters/item.rb]: Is there a reason for using the accessors here, rather than assigning to an ivar as elsewhere?

File: app/presenters/publishing_api_presenters/unpublishing.rb [hunk 14]
```diff
@@ -1,87 +1,76 @@
...
+      full_uri = Addressable::URI.parse(item.alternative_url)
 
-  def details
-    {
-      explanation: unpublishing_explanation,
-      unpublished_at: unpublishing.created_at,
-      alternative_url: unpublishing.alternative_url
-    }
-  end
+      path_uri = Addressable::URI.new
```
Comment (benlovell) [inline app/presenters/publishing_api_presenters/unpublishing.rb]: `new` takes an options hash including `path`, `query`, `fragment` and friends. That might tidy this up a teeny bit.

File: test/unit/presenters/publishing_api_presenters/edition_test.rb [hunk 15]
```diff
@@ -136,27 +135,27 @@ def present(edition, options={})
     stub_content_register_policies
     edition = create(:publication, :published,
```
Comment (benlovell) [inline test/unit/presenters/publishing_api_presenters/edition_test.rb]: I think it's clearer and (at least for me, anyway) faster to parse multiline arguments like this:

```ruby
create(
  :publication,
  :published,
  policy_content_ids: #...
)
```

File: app/presenters/publishing_api_presenters/item.rb (hunk 13 above)
Comment (boffbowsh) [inline app/presenters/publishing_api_presenters/item.rb]: Only preference.

Comment (benlovell) [inline app/presenters/publishing_api_presenters/item.rb]: :+1:

File: app/presenters/publishing_api_presenters/unpublishing.rb (hunk 14 above)
Comment (boffbowsh) [inline app/presenters/publishing_api_presenters/unpublishing.rb]: Updated

File: test/unit/presenters/publishing_api_presenters/edition_test.rb (hunk 15 above)
Comment (boffbowsh) [inline test/unit/presenters/publishing_api_presenters/edition_test.rb]: Done

**Comment by jamiecobbett** [comment]
This looks ready to merge (once green) :sweat: :fireworks:
